### PR TITLE
ts-migration/no-hooks

### PR DESCRIPTION
**Pull request by G-Rath**
It might be best to rebase `reapply-ts` on `master`, then rebase this PR 😬 

@SimenB can I leave that to you?

### CODE REVIEW

**Comment by SimenB** [comment]
>@SimenB can I leave that to you?

fo sho

File: src/rules/no-hooks.ts [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+    type: 'suggestion',
+  },
+  defaultOptions: [{ allow: [] } as { allow: readonly HookName[] }],
+  create(context, [{ allow }]) {
+    const whitelistedHookNames = allow.reduce((hashMap, value) => {
+      hashMap[value] = true;
+      return hashMap;
+    }, Object.create(null));
+
+    const isWhitelisted = (node: { callee: { name: string } }) =>
```
Comment (SimenB) [inline src/rules/no-hooks.ts]:
```suggestion
    const isWhitelisted = (node: JestFunctionCallExpressionWithIdentifierCallee<HookName>) =>
```

Comment (SimenB) [reply src/rules/no-hooks.ts]: Could we possibly do `export type JestHookFunction = JestFunctionCallExpressionWithIdentifierCallee<HookName>` or something? 🙂

Comment (G-Rath) [reply src/rules/no-hooks.ts]: > Could we possibly do...

fo sho

>The name is so long... 😛

Yeah, it is - I might do a clean up PR soon, and rename them to something like `Fn` to make them a touch shorter.

Sadly, that's just pretty much what always happens when working w/ AST nodes - there's so much contextual type information that you end up w/ really very verbose but long type names :joy:

Comment (SimenB) [reply src/rules/no-hooks.ts]: Yeah, we can go through after everything is migrated. I bet we'll see some patterns or something we can abstract out to make it a bit more readable at that point

**Pull request merged**